File: rlcore/relink_relay.cpp
// Standalone UDP relay daemon: forwards ReLink data frames between
// clients that registered for the same topic, for use when direct
// peer-to-peer NAT hole punching cannot cross a NAT/firewall at all
// (see relay_wire.hpp and README Step 13 for why a relay works where
// punching structurally can't -- both clients only ever open a NAT
// mapping toward this daemon's one fixed address, never toward each
// other).
//
// Performance: single recv buffer per iteration, reused for every
// forward -- a data frame is never re-encoded or copied into a second
// buffer, only sent on (sendto with the same pointer/length recvfrom
// just filled in). No heap allocation on the forwarding path: the
// per-topic member list is a fixed-capacity array reused in place, and
// membership lookups are pointer/index operations only.
//
// usage: relink-relay [port]
#include "relink/relay_wire.hpp"
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

using namespace relink;

namespace {

// A member's endpoint plus when it last registered/kept alive, so
// stale entries (crashed/departed clients) get pruned instead of
// accumulating forever.
struct Member {
    struct sockaddr_in addr;
    time_t last_seen;
};

constexpr time_t kMemberTtlSeconds = 30; // must outlive the client's re-register interval

bool same_addr(const struct sockaddr_in& a, const struct sockaddr_in& b) {
    return a.sin_addr.s_addr == b.sin_addr.s_addr && a.sin_port == b.sin_port;
}

} // namespace

int main(int argc, char** argv) {
    uint16_t port = (argc > 1) ? static_cast<uint16_t>(std::atoi(argv[1])) : kRelayDefaultPort;

    int sock = ::socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) {
        std::perror("socket");
        return 1;
    }

    struct sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = INADDR_ANY;
    addr.sin_port = htons(port);
    if (::bind(sock, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)) < 0) {
        std::perror("bind");
        return 1;
    }

    std::printf("relink-relay listening on 0.0.0.0:%u\n", port);
    std::fflush(stdout);

    std::unordered_map<uint32_t, std::vector<Member>> groups;
    uint8_t buf[kMaxFrameBytes];
    time_t last_sweep = std::time(nullptr);

    for (;;) {
        struct sockaddr_in src{};
        socklen_t src_len = sizeof(src);
        ssize_t n = ::recvfrom(sock, buf, sizeof(buf), 0,
                                reinterpret_cast<struct sockaddr*>(&src), &src_len);
        if (n <= 0) continue;

        uint32_t topic_id = 0;
        time_t now = std::time(nullptr);

        if (decode_relay_register(buf, static_cast<size_t>(n), &topic_id)) {
            auto& members = groups[topic_id];
            bool found = false;
            for (auto& m : members) {
                if (same_addr(m.addr, src)) { m.last_seen = now; found = true; break; }
            }
            if (!found) members.push_back(Member{src, now});
        } else if (peek_frame_topic_id(buf, static_cast<size_t>(n), &topic_id)) {
            auto it = groups.find(topic_id);
            if (it != groups.end()) {
                for (const auto& m : it->second) {
                    if (same_addr(m.addr, src)) continue; // never echo back to the sender
                    ::sendto(sock, buf, static_cast<size_t>(n), 0,
                             reinterpret_cast<const struct sockaddr*>(&m.addr), sizeof(m.addr));
                }
            }
        }
        // Anything else (malformed/unrecognized) is silently dropped --
        // a relay must never misinterpret bytes it can't identify.

        // Sweep expired members roughly once a second, not on every
        // packet -- keeps the hot path free of a per-packet time() call
        // beyond the one already taken above for last_seen bookkeeping.
        if (now != last_sweep) {
            last_sweep = now;
            for (auto& kv : groups) {
                auto& members = kv.second;
                members.erase(
                    std::remove_if(members.begin(), members.end(),
                                    [&](const Member& m) { return now - m.last_seen > kMemberTtlSeconds; }),
                    members.end());
            }
        }
    }
}

File: python/rlcore/relink_relay.py
#!/usr/bin/env python3
"""Standalone UDP relay daemon (Python) -- companion to the C++
relink-relay for environments without a compiler. Forwards ReLink data
frames between clients registered for the same topic; see relink.hpp's
relay_wire.hpp and README Step 13 for why a relay reaches nodes that
direct peer-to-peer NAT punching structurally cannot.

Performance notes (the honest ceiling for pure-Python UDP forwarding,
no C extension): one recvfrom per iteration into a bytes object the
socket layer already had to allocate -- there's no way to avoid that
allocation from pure Python, unlike the C++ relay's reused stack
buffer. What IS avoided: no re-encoding of the frame (the exact bytes
recvfrom returned are handed straight to sendto for every destination,
never rebuilt), no per-packet topic lookup beyond one dict get, and no
per-packet time.time() call for TTL sweeping (that's batched, not
inline -- only the register/keepalive path pays for one).

usage: relink_relay.py [port]
"""
import socket
import sys
import time

sys.path.insert(0, __file__.rsplit("/", 2)[0] + "/relink_py")
from relink.relay_wire import RELAY_DEFAULT_PORT, decode_relay_register, peek_frame_topic_id

MEMBER_TTL_SECONDS = 30  # must outlive the client's re-register interval


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else RELAY_DEFAULT_PORT

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(("0.0.0.0", port))
    print(f"relink-relay (python) listening on 0.0.0.0:{port}", flush=True)

    groups = {}  # topic_id -> {addr: last_seen}
    last_sweep = time.time()

    while True:
        try:
            data, addr = sock.recvfrom(1500)
        except OSError:
            continue
        now = time.time()

        topic_id = decode_relay_register(data)
        if topic_id is not None:
            groups.setdefault(topic_id, {})[addr] = now
        else:
            topic_id = peek_frame_topic_id(data)
            if topic_id is not None:
                members = groups.get(topic_id)
                if members:
                    for member_addr in members:
                        if member_addr == addr:
                            continue  # never echo back to the sender
                        try:
                            sock.sendto(data, member_addr)
                        except OSError:
                            pass
            # else: unrecognized bytes -- silently dropped, a relay must
            # never misinterpret bytes it can't identify.

        if now - last_sweep >= 1.0:
            last_sweep = now
            for topic_id in list(groups.keys()):
                members = groups[topic_id]
                stale = [a for a, seen in members.items() if now - seen > MEMBER_TTL_SECONDS]
                for a in stale:
                    del members[a]
                if not members:
                    del groups[topic_id]


if __name__ == "__main__":
    main()
